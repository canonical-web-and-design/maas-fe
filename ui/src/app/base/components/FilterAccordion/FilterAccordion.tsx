import type { ReactNode } from "react";
import { useMemo, useState } from "react";

import {
  Accordion,
  Button,
  ContextualMenu,
  List,
} from "@canonical/react-components";
import classNames from "classnames";

import type { FilterValue } from "app/utils/search/filter-handlers";
import type FilterItems from "app/utils/search/filter-items";

// The key for the filter, this will usually be a model attribute.
type FilterKey = string;

// A mapping between the available values for a filter and the count of items
// that have that value.
type FilterValues = Map<FilterValue, number>;

// A mapping between filters and the available values and counts.
type FilterSections = Map<FilterKey, FilterValues>;

export type Props<I, PK extends keyof I> = {
  filterItems: FilterItems<I, PK>;
  filterNames: Map<FilterKey, string>;
  filterOrder: FilterKey[];
  filterString?: string;
  getValue: (item: I, filter: FilterKey) => FilterValue | FilterValue[] | null;
  getValueDisplay?: (filter: FilterKey, value: FilterValue) => ReactNode;
  items: I[];
  onUpdateFilterString: (filterString: string) => void;
};

// An accordion section.
type Section = {
  title?: string;
  content: ReactNode;
  key: string;
};

const getFilters = <I, PK extends keyof I>(
  items: Props<I, PK>["items"],
  filterOrder: Props<I, PK>["filterOrder"],
  getValue: Props<I, PK>["getValue"]
) => {
  const filters: FilterSections = new Map();
  items.forEach((item) => {
    filterOrder.forEach((filter) => {
      const value = getValue(item, filter);
      // Ignore this section if this is not a value that can be filtered by:
      if (!value) {
        return;
      }
      // Turn everything into an array so we can loop over all values.
      const valueArray = Array.isArray(value) ? value : [value];
      // Remove any values that are not useful for filtering, e.g. null/boolean/undefined.
      valueArray.filter((filterValue) => Boolean(filterValue));
      // Ignore this section if there are no useful values to filter by:
      if (valueArray.length === 0) {
        return;
      }
      // If this section does not already exist then create a new map to store
      // the values and counts.
      let storedFilter = filters.get(filter);
      if (!storedFilter) {
        const filterValues: FilterValues = new Map();
        filters.set(filter, filterValues);
        storedFilter = filters.get(filter);
      }
      valueArray.forEach((filterValue) => {
        if (storedFilter) {
          const storedValue = storedFilter.get(filterValue) || 0;
          storedFilter.set(filterValue, storedValue + 1);
        }
      });
    });
  });
  return filters;
};

const sortByFilterKey = (
  a: [FilterValue, number],
  b: [FilterValue, number]
) => {
  const aValue = a[0];
  const bValue = b[0];
  if (aValue < bValue) {
    return -1;
  }
  if (aValue > bValue) {
    return 1;
  }
  return 0;
};

const FilterAccordion = <I, PK extends keyof I>({
  filterItems,
  filterNames,
  filterOrder,
  filterString,
  getValue,
  getValueDisplay,
  items,
  onUpdateFilterString,
}: Props<I, PK>): JSX.Element => {
  const currentFilters = filterItems.getCurrentFilters(filterString);
  const [expandedSection, setExpandedSection] = useState();
  const sections = useMemo(() => {
    const filterOptions = getFilters<I, PK>(items, filterOrder, getValue);
    return filterOrder.reduce<Section[]>((options, filter) => {
      const filterValues = filterOptions.get(filter);
      if (filterValues && filterValues.size > 0) {
        options.push({
          title: filterNames.get(filter),
          content: (
            <List
              className="u-no-margin--bottom"
              items={Array.from(filterValues)
                .sort(sortByFilterKey)
                .map(([filterValue, count]) => (
                  <Button
                    appearance="base"
                    className={classNames(
                      "u-align-text--left u-no-margin--bottom filter-accordion__item is-dense",
                      {
                        "is-active": filterItems.isFilterActive(
                          currentFilters,
                          filter,
                          filterValue,
                          true
                        ),
                      }
                    )}
                    data-test={`filter-${filter}`}
                    onClick={() => {
                      const newFilters = filterItems.toggleFilter(
                        currentFilters,
                        filter,
                        filterValue,
                        true
                      );
                      onUpdateFilterString(
                        filterItems.filtersToString(newFilters)
                      );
                    }}
                  >
                    {getValueDisplay
                      ? getValueDisplay(filter, filterValue)
                      : filterValue}{" "}
                    ({count})
                  </Button>
                ))}
            />
          ),
          key: filter,
        });
      }
      return options;
    }, []);
  }, [
    currentFilters,
    filterItems,
    filterNames,
    filterOrder,
    getValue,
    getValueDisplay,
    items,
    onUpdateFilterString,
  ]);

  return (
    <ContextualMenu
      className="filter-accordion"
      constrainPanelWidth
      hasToggleIcon
      position="left"
      toggleClassName="filter-accordion__toggle"
      toggleLabel="Filters"
    >
      <Accordion
        className="filter-accordion__dropdown"
        expanded={expandedSection}
        externallyControlled
        onExpandedChange={setExpandedSection}
        sections={sections}
      />
    </ContextualMenu>
  );
};

export default FilterAccordion;
