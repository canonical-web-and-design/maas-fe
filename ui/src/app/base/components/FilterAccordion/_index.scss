@mixin FilterAccordion {
  [class^="p-button"].has-icon.filter-accordion__toggle,
  .filter-accordion__item,
  .filter-accordion {
    width: 100%;
  }

  [class^="p-button"].has-icon.filter-accordion__toggle {
    padding-left: $spv-inner--small;
    text-align: left;

    i {
      position: absolute;
      right: $spv-inner--large;
      top: calc(#{$spv-inner--medium} - 1px);
    }
  }

  .filter-accordion .p-accordion__list {
    margin-bottom: 0;
  }

  .filter-accordion__dropdown {
    max-height: 66vh;
    overflow-y: auto;
  }

  .filter-accordion .p-accordion__panel {
    padding-left: 0;
  }

  .filter-accordion .p-list__item {
    padding-bottom: 0;
    padding-top: 0;
  }

  .filter-accordion .p-contextual-menu__dropdown {
    max-width: none;
    width: calc(100% - 1px);
  }

  .filter-accordion__item {
    padding-left: $sp-unit * 5;

    &.is-active {
      background-image: url("data:image/svg+xml,%3Csvg width='22' height='16' xmlns='http://www.w3.org/2000/svg'%3E%3Cg fill='none' fill-rule='evenodd'%3E%3Cpath d='M-1-1h24v24H-1z'/%3E%3Cpath fill='" + vf-url-friendly-color(
          $color-mid-dark
        ) + "' fill-rule='nonzero' d='M2.872 5.933L.6 8.205l7.733 7.734L21.4 2.872 19.128.6 8.333 11.397z'/%3E%3C/g%3E%3C/svg%3E");
      background-position-x: $sph-inner;
      background-position-y: center;
      background-repeat: no-repeat;
      background-size: map-get($icon-sizes, default);
      font-weight: 400;
    }
  }
}
