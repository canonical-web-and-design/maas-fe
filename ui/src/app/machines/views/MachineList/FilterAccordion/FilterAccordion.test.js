import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import React from "react";

import FilterAccordion from "./FilterAccordion";

const mockStore = configureStore();

describe("FilterAccordion", () => {
  let state;
  beforeEach(() => {
    state = {
      config: {
        items: []
      },
      machine: {
        errors: {},
        loading: false,
        loaded: true,
        items: [
          {
            link_speeds: [100],
            pool: {
              id: 1,
              name: "pool1"
            },
            zone: {
              id: 1,
              name: "zone1"
            }
          }
        ]
      }
    };
  });

  it("renders", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FilterAccordion setSearchText={jest.fn()} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("FilterAccordion")).toMatchSnapshot();
  });

  it("displays a spinner when loading machines", () => {
    state.machine.loaded = false;
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FilterAccordion setSearchText={jest.fn()} />
        </MemoryRouter>
      </Provider>
    );
    // Open the menu:
    wrapper.find("Button.p-contextual-menu__toggle").simulate("click");
    expect(wrapper.find("Loader").exists()).toBe(true);
  });

  it("formats link speeds", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FilterAccordion setSearchText={jest.fn()} />
        </MemoryRouter>
      </Provider>
    );
    // Open the menu:
    wrapper.find("Button.p-contextual-menu__toggle").simulate("click");
    expect(
      wrapper
        .find(".filter-accordion__item")
        .findWhere(button => button.text().includes("100 Mbps"))
        .exists()
    ).toBe(true);
  });

  it("can mark an item as active", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FilterAccordion
            searchText="pool:(=pool1)"
            setSearchText={jest.fn()}
          />
        </MemoryRouter>
      </Provider>
    );
    // Open the menu:
    wrapper.find("Button.p-contextual-menu__toggle").simulate("click");
    expect(wrapper.find(".filter-accordion__item.is-active").exists()).toBe(
      true
    );
  });

  it("can set a new filter", () => {
    const setSearchText = jest.fn();
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FilterAccordion setSearchText={setSearchText} />
        </MemoryRouter>
      </Provider>
    );
    // Open the menu:
    wrapper.find("Button.p-contextual-menu__toggle").simulate("click");
    wrapper
      .find(".filter-accordion__item")
      .at(0)
      .simulate("click");
    expect(setSearchText).toHaveBeenCalledWith("pool:(=pool1)");
  });

  it("hides filters if there are no values", () => {
    delete state.machine.items[0].link_speeds;
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FilterAccordion setSearchText={jest.fn()} />
        </MemoryRouter>
      </Provider>
    );
    // Open the menu:
    wrapper.find("Button.p-contextual-menu__toggle").simulate("click");
    expect(
      wrapper
        .find(".p-accordion__tab")
        .findWhere(button => button.text() === "Link speed")
        .exists()
    ).toBe(false);
  });

  it("hides filters if the value is an empty array", () => {
    state.machine.items[0].link_speeds = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <FilterAccordion setSearchText={jest.fn()} />
        </MemoryRouter>
      </Provider>
    );
    // Open the menu:
    wrapper.find("Button.p-contextual-menu__toggle").simulate("click");
    expect(
      wrapper
        .find(".p-accordion__tab")
        .findWhere(button => button.text() === "Link speed")
        .exists()
    ).toBe(false);
  });
});
