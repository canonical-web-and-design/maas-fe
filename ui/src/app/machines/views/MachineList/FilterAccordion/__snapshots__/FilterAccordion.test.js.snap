// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterAccordion renders 1`] = `
<FilterAccordion
  setSearchText={[MockFunction]}
>
  <ContextualMenu
    className="filter-accordion"
    constrainPanelWidth={true}
    dropdownContent={
      <Accordion
        className="filter-accordion__dropdown"
        sections={
          Array [
            Object {
              "content": <List
                className="u-no-margin--bottom"
                items={
                  Array [
                    <Button
                      appearance="base"
                      className="u-align-text--left u-no-margin--bottom filter-accordion__item is-dense"
                      onClick={[Function]}
                    >
                      pool1
                       
                      (
                      1
                      )
                    </Button>,
                  ]
                }
              />,
              "title": "Resource pool",
            },
            Object {
              "content": <List
                className="u-no-margin--bottom"
                items={
                  Array [
                    <Button
                      appearance="base"
                      className="u-align-text--left u-no-margin--bottom filter-accordion__item is-dense"
                      onClick={[Function]}
                    >
                      zone1
                       
                      (
                      1
                      )
                    </Button>,
                  ]
                }
              />,
              "title": "Zone",
            },
            Object {
              "content": <List
                className="u-no-margin--bottom"
                items={
                  Array [
                    <Button
                      appearance="base"
                      className="u-align-text--left u-no-margin--bottom filter-accordion__item is-dense"
                      onClick={[Function]}
                    >
                      undefined nodes
                       
                      (
                      1
                      )
                    </Button>,
                  ]
                }
              />,
              "title": "NUMA nodes",
            },
            Object {
              "content": <List
                className="u-no-margin--bottom"
                items={
                  Array [
                    <Button
                      appearance="base"
                      className="u-align-text--left u-no-margin--bottom filter-accordion__item is-dense"
                      onClick={[Function]}
                    >
                      Not supported
                       
                      (
                      1
                      )
                    </Button>,
                  ]
                }
              />,
              "title": "SR-IOV support",
            },
            Object {
              "content": <List
                className="u-no-margin--bottom"
                items={
                  Array [
                    <Button
                      appearance="base"
                      className="u-align-text--left u-no-margin--bottom filter-accordion__item is-dense"
                      onClick={[Function]}
                    >
                      100 Mbps
                       
                      (
                      1
                      )
                    </Button>,
                  ]
                }
              />,
              "title": "Link speed",
            },
          ]
        }
      />
    }
    hasToggleIcon={true}
    position="left"
    toggleClassName="filter-accordion__toggle"
    toggleLabel="Filters"
  >
    <span
      className="filter-accordion p-contextual-menu--left"
    >
      <Button
        aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
        aria-expanded="false"
        aria-haspopup="true"
        className="p-contextual-menu__toggle filter-accordion__toggle"
        hasIcon={true}
        onClick={[Function]}
      >
        <button
          aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
          aria-expanded="false"
          aria-haspopup="true"
          className="p-button--neutral has-icon p-contextual-menu__toggle filter-accordion__toggle"
          onClick={[Function]}
        >
          <span>
            Filters
          </span>
          <i
            className="p-icon--contextual-menu p-contextual-menu__indicator"
          />
        </button>
      </Button>
    </span>
  </ContextualMenu>
</FilterAccordion>
`;
